### DISCUSSION TITLE
Support for Hauwei EMMA-A01/A02 (New Product Release)

### DISCUSSION THREAD

**Discussion by Roving-Ronin** [wlcrs/huawei_solar#457] [Ideas]
FYI giving the release of the EMMA in Q4 2023, the DTSU-666H, SmartLogger and Dongle look to be superceded as the "default" connectivity option for Huawei inverters/batteries etc.  Given this thought it might be an idea to open a discussion upon this new product, even though pending its release currently, to allow planning / looking into it for the future.

Read below and you can work out that EMMA is collapsing the Logger functionality, the meters functions (with its 63A CT sensors, for home installs given 63A that is same as the BackupBox and its replacement the 'SmartGuard') and the network connectivity all into one 6 DIN unit. Atop of this it will also provide some additional smarts / functions.

Going forward this raises the question will WLCRS integration support this hardware (i.e. uses the same registers as now), OR will it use different registers etc and need updates in the WLCRS integration to support it? At least with Huawei merging all these functions into one bit of kit going forward, this will also mean one standard method of connectivity (at least for home installations) making support / troubleshooting etc easier. 

IMHO verses all the different connectivity options (Dongle, Elfin or SmartLogger [that doesn't 100% work]) this could be a positive development and well worth upgrading from the DDSU-666H / DTSU-666H and dongle to this 'all in one' solution, whilst also adding AI/Cloud based data for automation and decision making.
____________________________
![image](https://github.com/wlcrs/huawei_solar/assets/108674933/bccd04d1-8c1f-4411-a42b-5e646d3ab12a)

EMMA-A01/A02 is Huawei's new energy management assistant, the ideal complement for a smart and efficient home with solar power systems with Huawei inverters and batteries.

Huawei EMMA-A01/A02 Power Management Assistant

With this AI-based power management assistant, Huawei provides power generation estimates based on big data such as historical data recorded by Huawei inverters or weather forecasts. Thanks to all this information, it adjusts the entire system to make it much more efficient and intelligent.

- Supports heat pump and smart circuit breaker for comprehensive management of home energy use.

- Supports 63A current measurement with internal CT for easy installation.

- Supports scheduled mode and PV array preferences mode for best electricity savings.
____________________________

Note: Trying to see if access can be gained to at least modbus register definitions for EMMA, to allow evaluation for WLCRS integration compability.

**Comment by wlcrs** [comment]
As a general policy this integration only supports devices I can test against (ie. the devices that I own myself). Only so can I guarantee the proper working of this integration. 

I do welcome contributions that add support for new devices. So feel free to create a PR in the underlying [`huawei-solar` library](https://gitlab.com/Emilv2/huawei-solar/). I would advice anyone contributing to create a PR early in their development work to get feedback and tips from me.

**Reply by Roving-Ronin** [reply]
This thread is to highlight:

A) There is a new product coming out, which, from information available will be the default method for Huawei installations to be deployed with.

B) Raises this for awareness and future planning for how can this be added/supported in the integration? (Do we need a GoFundMe to get WLCRS's setup upgraded from a dongle to EMMA in future?).

C) Highlights that, with information being that it will replace the DDSU/DTSU-666H meter, Dongle and SmartLoggers, it would not be worth pursuing support for SmartLoggers (at least for residential users of them). Instead of Huawei is standardising upon the EMMA going forward, it would be easier to just upgrade a SmartLogger to an EMMA setup.

D) Provides a discussion thread for people to follow and communicate upon the EMMA and if/how to support it.

FYI also trying to see if Huawei will release 'Modbus Register Definitions' documentation of the EMMA, allowing a paper evaluation of if the integration will work with it (same registers used), or if anything is different from existing registers.

**Comment by Roving-Ronin** [comment]
Links to installation and user guides for EMMA, and mention of how it connects up, posted in Issues section.
See: [Feature Request]: Wallbox Integration - https://github.com/wlcrs/huawei_solar/issues/491 

Edit: I should add, information received is that EMMA will support connecting to cloud services with the example provided being that it will be able to check for environmental / weather data (no info as to which clould service provider) and based upon this make intelligent decisions, such as ' rain / overcast day tomorrow with low expected kWh to be generated via local PC, then instruct LUNA ESS to charge overnight during offpeak when cost per kWh is lowest'.

**Comment by tmycgn** [comment]
Does anyone know whether the Emma and the dongle can be operated in parallel? I would like to use this integration. Otherwise, I agree that all new installations will only be delivered with "new" Emma. I would therefore also be very happy about an integration of Emma. :)

**Reply by Roving-Ronin** [reply]
@tmycgn 
EMMA replaces the sDongle in an installation, as only ONE device (made to provide the internet connection) can exist in a plant  (be it 1 or 3 invertes cascaded), hence logic dictates that its would not be compatible with running EMMA + sDongle concurrently.

Have you tried installing the WLCRS integration using the modbus id # (1?) of the EMMA on your LAN, with modbus support activated?
If it doesn't work, your only options currently would be to provide Thijs (WLCRS) access to your HA to develop support for EMMA upon your HA+Huawei instance. The only other fallback that you (may) be able to get working, would mean needing to get something like the Elfin (see Wiki) wired in upstream of the Primary Inverter (or EMMA?) and seeing they both still work (i.e. can access mmodbus and EMMA can talk to FusionSolar portal).

Re new instalations, the sDongle would appear to be remaining as a connectvitiy option going forward, but it also has its own issues such as if people install a Huawei EV charger and connect it to the modbus on the sDongle, that will boot HA from connecting (as it only supports 1 connection). On the flip side, people in future installing the 'Backup Box' will be looking at the replacement the 'SmartGuard' that includes an EMMA and DDSU/DSTU type consumption meter built into it (so they won't have any choice).

As Thijs (WCLRS) comments, as it stands he can't add support for anything until has can get access to this new Huawei equipment that is coming to market (i.e. either people give him remote access to develop, or GoFundMe required).

**Reply by wlcrs** [reply]
For the moment, Huawei is not able to produce any documentation on how to connect to installations with an EMMA device present. After asking for documentation, they first erroneously sent docs concerning SDongleA devices. After a follow up email pressing for docs for connecting to EMMA via Modbus, we got the following answer:

> Unfortunately we do not have any ModBus guidance especially for EMMA-A02.

A second follow up did not result in a more promising answer:

> Unfortunately, we do not have any guideline about how access works exactly and how the individual services can be accessed (modbus tCP, MQTT, web,...) especially for EMMA-A02.

I've already had access to an installation that has an EMMA device present, but was not able to retrieve any information from it over Modbus (RTU or TCP). The device did not answer to any Modbus request.

As it stands now, I cannot confirm if this integration will ever work with such installations.

More info in #498

**Reply by tmycgn** [reply]
Thanks for the information @Roving-Ronin, that saved me a failed invest (sDongle). :) No, I haven't tried it yet as I don't have the system yet. The reason was to verify it in advance. But if I will try it and post it here. ✌️

**Reply by Roving-Ronin** [reply]
@tmycgn as in you don't have an EMMA, or don't have the Huawei inverter yet? If you haven't got the inverter yet, IF your a 3 phase user (i.e. getting say an M1 inverter), then it wouldn't be a failed investment as some models of inverters (such as the M1) actually come bundled with an sDongle from the factory. Alas 'some' installers like to pocket them / add them to the quote as an optional item for additional cost (cost AUD$99 wholesale typically).

**Reply by abrandl70** [reply]
Not sure if this is old news, but I can confirm that EMMA and Dongle don’t work together. I had a new installation 3 weeks ago and the EMMA only found the other devices after the Dongle was removed from the installation.

**Comment by SwissDani** [comment]
I have installed an Emma A02. With the newest update (V100R024C00SPC100) dated 15.07.2024 I got a pdf for the MODBUS Interface Defintions. Maybe this can be used for the integration?
I have no experience with MODBUS. But I am an experienced .NET developer and would be happy to help / support with tests.

Additional Infos form the new Communication Matrix:
![image](https://github.com/user-attachments/assets/1a91ee6f-f607-44c9-9604-0935dbab930b)


[SmartHEMS V100R024C00 MODBUS Interface Definitions.pdf](https://github.com/user-attachments/files/16286512/SmartHEMS.V100R024C00.MODBUS.Interface.Definitions.pdf)

**Reply by tmycgn** [reply]
👍 1-2 weeks ago I had written to the support and request for forwarding to the product team asking for support with the Modbus interface with reference to this thread. Maybe we could move something. :) I have looked at the document, but unfortunately it is also a bit too techy for me, so I am hoping for support here.

**Reply by StefanOltmann** [reply]
Finally!

Unfortunatenly, the FusionSolar app does not find the update package yet.

[Direct download](https://support.huawei.com/enterprise/en/fusionsolar/smarthems-pid-256539983/software/262461879?idAbsPath=fixnode01%7C258788303%7C258788491%7C258789989%7C23205712%7C256539983) requires a login.

**Reply by tmenguy** [reply]
Great … so there is hope that Emma could work with HA (will wait to get one for the new and way better backup box then …)

**Reply by Roving-Ronin** [reply]
'Get FOR'... fyi the SmartGuard (Backup Box) includes an EMMA. Top left side of its internal layout.

**Reply by tmycgn** [reply]
The EMMA-A02 is integrated in the new backup box that will be released later this year. I think and hope that there will be an upgrade package for customers who already own the Emma so that they don't have to buy it twice.

**Reply by StefanOltmann** [reply]
Yes, let's hope that.

**Reply by tmycgn** [reply]
See here: https://solar.huawei.com/en/professionals/all-products/HUAWEI-SmartGuard

<img width="837" alt="image" src="https://github.com/user-attachments/assets/0b851ec7-e4c3-4e60-a5c6-e079acf652a2">

**Reply by tmenguy** [reply]
Yep it is the one I was looking for (in 3 phases) … but no way if no HA :)

**Reply by StefanOltmann** [reply]
The question is: With the downloaded ZIP file on my computer, how can I install that firmware on my EMMA using my mobile phone? Do I need to wait until it shows up on the phone?

**Reply by bepeearth** [reply]
Hey, yes, I was able to install the update described there and provided to me by support without any errors. However, when retrieving Modbus TCP/IP with Nodered via port 502, I always only get the value “zero” under the register addresses.

Something doesn't seem to be working, or maybe something isn't activated yet!!

**Reply by SwissDani** [reply]
I had to activate it with the app as an installer.
But didnt had time to test it
![Screenshot_20240719_180711_FusionSolar](https://github.com/user-attachments/assets/8b1f6bac-77d7-437e-87d4-ef78f40a2a51)

**Reply by StefanOltmann** [reply]
I don’t have installer access to FusionSolar Web unfortunately. I’m an end user.
But using the mobile app I can enter installer mode as the default password was set.

So I need to wait?

**Reply by SwissDani** [reply]
I used the App not the Web. In the app i used "Inbetriebnahme des Gerät", connected to local Emma Wifi and used the Installer Login.

**Reply by StefanOltmann** [reply]
Yes, but „Dateidownload“ says „Die Datei ist auf dem neusten Stand“.
I don’t see where I can select the downloaded firmware ZIP from disc.

**Reply by ismiregal17** [reply]
Hello, thanks for the tip. Modbus TCP is now activated but unfortunately there is no connection via WLAN. What is the client ip address

![IMG_4103](https://github.com/user-attachments/assets/385d358e-f965-4d8a-8926-02324e03f55c)

**Reply by SwissDani** [reply]
@StefanOltmann I was also unable to install the update myself via the app. I was able to select the zip file but it then reported that the version was incorrect. I therefore had the update installed via an installer using the web portal.
I was then able to activate access myself via local WLAN and the app

@ismiregal17 I have also tested the MODBUS access with various tools and settings but it did not work.

Has anyone already been able to successfully establish a connection?

**Reply by ismiregal17** [reply]
@SwissDani No, unfortunately it doesn't work yet. And the bugs are still present in the Fusionsolar app and in the portal. Display in gigawatts instead of kilowatts and therefore the EMMA function is not possible or only possible to a limited extent.

**Reply by wlcrs** [reply]
Interesting to see that Huawei finally released documentation for the EMMA device!

For those wanting to test: 

1. Please disable the TCP encryption toggle for now:

![image](https://github.com/user-attachments/assets/30c2b384-fda4-4411-9eba-086a1bae1610)

2, You should not use slave 0, as the integration does not have any detection logic, nor support for the limited number of registers that the EMMA device exposes. Only put the slave numbers of your inverters in the list during intial configuration.

Can somebody take a sceenshot of what is behind that "Modbus TCP certificate" button?

**Reply by ismiregal17** [reply]
![IMG_4104](https://github.com/user-attachments/assets/ec0c45ce-6e4e-490d-917a-7dc6088aa82e)

**Reply by StefanOltmann** [reply]
So you all asked your installer to install the update using the web for you?

it doesn’t make sense to me that I can update the devices using my phone but not using the web portal as user.

**Reply by Roving-Ronin** [reply]
Hi Thijs,

In the Huawei Library repo, what would be the best way to start contributing by adding the sensors etc for EMMA? 
Instead of adding into the Main branch would it be better if a EMMA branch was created for now, that when all completed you can then merge into Main branch?

Also I note that some sensors use the same register as the existing sensors, albeit some in the Huawei documentation are called by a different name. Should these registers be added into the library, under the stub you've already made for EMMA, or just the new registers that EMMA provides ?

**Reply by wlcrs** [reply]
Development should start in the underlying library first. Feel free to create a PR in https://gitlab.com/Emilv2/huawei-solar . Please have a look at the last commit in which I did some preliminary work to provide a guide on how the registers can be added and be properly tagged: https://gitlab.com/Emilv2/huawei-solar/-/blob/feature/emma/src/huawei_solar/registers.py#L742-746

**Reply by wlcrs** [reply]
@ismiregal17 thank you for that screenshot. It shows that Huawei has implemented proper Modbus TLS support with the ability to set your own root and client certificates. Getting this setup is not straightforward if you're not familiar with generating, signing and renewing X.509 certificates. I don't plan on adding support for this, as there is little added value for having encrypted communication in your home network. Luckily, you can just disable it with a toggle.

**Reply by ismiregal17** [reply]
@wlcrs hello, thank you for your reply. It would be really interesting for me if the EMMA can then be integrated in HA with Huawei Solar. And if so, how I have to carry out the integration. I find your repository perfect for Huawei and would like to work with it again. 
thank you and king regards

**Reply by ismiregal17** [reply]
Here are the certificates with the Huawei description
![IMG_4114](https://github.com/user-attachments/assets/5d6a7afb-229a-4b19-88a3-d720968dc810)

**Reply by tmycgn** [reply]
If I would try to create a comparison of the registers between SUN2000 and EMMA, could someone provide me with the link to the correct comparison document of the SUN2000? I have already found this one, would this be the right one? https://community.symcon.de/uploads/short-url/pqZXWOienoBK2AsEzGD2oH1bcPR.pdf

[Solar Inverter Modbus Interface Definitions (V3.0).pdf](https://github.com/user-attachments/files/16329209/Solar.Inverter.Modbus.Interface.Definitions.V3.0.pdf)

**Reply by ismiregal17** [reply]
This is the right document

**Reply by Roving-Ronin** [reply]
Modbus definitions get updated periodically and the most up to date documentation is in the firmware/SPC 'Release Document (zip)'.

**Latest for SUN2000MA - L0 / L1 / M0 / M1 etc:**
[SUN2000MA Solar Inverter Modbus Interface Definitions - 2024-07-04 v08 SPC165.pdf](https://github.com/user-attachments/files/16334790/SUN2000MA.Solar.Inverter.Modbus.Interface.Definitions.-.2024-07-04.v08.SPC165.pdf)
Note: This document also contains the definitions for the LUNA ESS and 'Meter Equipment' (i.e. DTSU/DDSU-666)

**Latest for SUN2000MB - MB0 / M5 :**
[SUN2000MB Solar Inverter Modbus Interface Definitions - 2024-07-15 v03 SPC204.pdf](https://github.com/user-attachments/files/16334813/SUN2000MB.Solar.Inverter.Modbus.Interface.Definitions.-.2024-07-15.v03.SPC204.pdf)

**Reply by tmycgn** [reply]
My approach: I am currently trying to build a GPT prompt that prepares the information for us. 🤖

**Reply by Roving-Ronin** [reply]
Thomas,  The interesting thing will be there's a number of EMMA sensors that are called something different, but when you look at them and the data option (i.e. say the sensor that defines in TOU with excess energy either push to battery or send to grid as the priority) for the sensor (especially some of the RW sensors) they are the same as some of the existing non-EMMA sensors, albeit by another name.

**Reply by tmycgn** [reply]
Hi, yes, I've seen that, technically I'm not quite as deep into it as you are: can you post any corrections I need here? Simply in the form: "IS", "SHOULD" string. I think I'm making little good progress, try to post a first result tonight (bot for creation) Excel with register comparison for the review. Hope that helps a bit.

**Reply by Roving-Ronin** [reply]
ATM I've only had time to list out the RO registers. Scratching my head re how to define the RW ones, let alone then needing to match them to existing registers/sensors. If they have the same register number then the existing entry can just be updated to add the target_device= option, however I have seen one or two already that actually have the same register number as an existing register, BUT are for a different function!.

RO registers, still needing filtering to see if existing can be used and also renaming to follow existing format (and add EMMA_ as a prefix) are:

Edit: <CODE REMOVED... PLEASE REFER TO LINK IN NEW POST BELOW>

**Reply by Roving-Ronin** [reply]
Not used to GitLab, so doing my updates in GitHub, from where I'll then paste them into GitLab. As it appears to be GitLab updates into the GitHub Huawei Library (not the other way around).

I'm done for the night, but there's some updated code at: https://github.com/Roving-Ronin/huawei-solar-lib/blob/dfd6f5dce7d20052c3a35e01994bfa2c1a866950/src/huawei_solar/registers.py#L1302. This supercedes what I posted in the above post (that I'll now remove from there).

Line 1302 to 1465. Please note that the registers where the rn. is NOT then followed by an EMMA_ prefix still have to be cross checked and see if there's existing registers (with the same number) that can be used. Also need to see if the name of them needs changing to align with Thijs's format (currently just taken verbatim from the EMMA doco).

There's some comments in there also, such as Line 1347, that I have no clue has to define 'no default value', and others where they have to have a CLASS defined in register_values.py. 


Thijs - Would you have an idea as to Section 3.3 'Register Definitions for a Charger' and how this should be defined? This appears to use some of the same existing register numbers, like 'Offering Name' (aka Model Name) on Register 30000.  Does this need a:

a) target_device=TargetDevice.EMMA_CHARGER    added?   and;
b) EMMA_CHARGER_REGISTERS = {   section created also ?

From reading the EMMA documentation and comparing to the last list of possible Huawei EV Charger modbus registers ( https://github.com/wlcrs/huawei_solar/issues/491#issuecomment-2094635359 ) this EMMA list looks like its a cut down list published via EMMA, in that is does't have any RW registers to allow setting Charge limits etc. So there would potentially still be a 'full blown' EVCharger list of registers to add seperately in future / for people not using an EMMA.

**Reply by tmenguy** [reply]
Hey, great and thx to start the mapping! ... but just to be sure, as of now no one has been able to connect to an EMMA, correct?

**Reply by tmycgn** [reply]
@Roving-Ronin I just asked the Huawei Support / Product Team again about the RW registers for the EV Charger via the same ticket I used on July 9th to request the creation of the Modbus definitions. Thanks for the hard work in adding the RO registers, the extraction via the 4o gpt bot failed after hours due to the amount of data. :/

**Reply by wlcrs** [reply]
I've implemented initial support. Please go to https://github.com/wlcrs/huawei_solar/issues/498#issuecomment-2266771661 for more information. 

This is untested, so I'm 90% certain that it will (partially) crash. If somebody is willing to provide me with remote access to an EMMA installation, then please give me a shout.

**Comment by Roving-Ronin** [comment]
Thijs,

Looking at the Huawei Library and in registers.py you've added some example EMMA entries, and updated some lines where EMMA and existing devices share the same register, i.e. 30000). I also note that a while ago you added a separate section for EMMA registers, at line: [1270](https://gitlab.com/Roving-Ronin/huawei-solar/-/blob/master/src/huawei_solar/registers.py?ref_type=heads#L1270)

Should we be placing registers that are shared/common between EMMA and existing devices in the section starting at https://gitlab.com/Roving-Ronin/huawei-solar/-/blob/master/src/huawei_solar/registers.py?ref_type=heads#L741
and EMMA only/specific registers in the EMMA section starting at line 1270?

If so, then https://gitlab.com/Roving-Ronin/huawei-solar/-/blob/master/src/huawei_solar/registers.py?ref_type=heads#L746 and https://gitlab.com/Roving-Ronin/huawei-solar/-/blob/master/src/huawei_solar/registers.py?ref_type=heads#L771 will also need to be moved the to EMMA specific section.

Also I note in the modbus register documentation reference to a 'U64' and 'I64' type of register, that appears to be new. Is this something that you are able to add support for (in ???).

Thanks.

**Reply by wlcrs** [reply]
Yes, move them to that EMMA section. I've added U64 and I64 register support on the `feature/emma` branch.

**Reply by wlcrs** [reply]
I've integrated a small part of your work for now. I think it is important to first validate that we are able to properly retrieve those first few registers before doing all the dull work of inputting in the registers.

**Reply by Roving-Ronin** [reply]
Hi Thijs,

Agree, as you've posted on the other threads, now we just need some of these EMMA owners to test and assist. 
PS, apologies been flat out last week or so with day to day work.
Let me know when/what is needed for doing some more of the 'dull' work, post user feedback.

On a side note, I think we are still waiting for one user to see if he has luck getting Modbus info for the EV charger. Given EMMA appears to have some registers for it, but they appear to only be a limited subset versus the expected list (or Emma ones dont include active power info or stop/start charge iirc).

**Comment by StefanOltmann** [comment]
I managed to update my EMMA to the recent update.

Did anyone manage to receive valid values so far?

If I run this script I always get "Slave Device Failure":
https://github.com/StefanOltmann/sun2000_modbus_test/blob/main/src/main/kotlin/Main.kt

Encryption is disabled, client IP is set to the computer executing the script.

**Reply by wlcrs** [reply]
Is your modbus slave_id set to 0 ? I assume the default of that library will be 1?

**Reply by StefanOltmann** [reply]
Oh, that was it. I totally missed that slave_id hint. Thank you a lot!

I receive valid values for the registers specified in the modbus documentation. Great.

**Reply by wlcrs** [reply]
Ok, great news! You verified that the EMMA device does react to Modbus TCP requests :-) Now it is up to me to get this integration working.

**Reply by StefanOltmann** [reply]
Sounds great! :)

Here is code that actually works with EMMA (modified from the previous script):
https://github.com/StefanOltmann/emma_modbus_test/blob/main/src/main/kotlin/Main.kt

The important parts are indeed `slave_id = 0` and setting the client IP address correctly.

The EMMA is significantly more stable than the dongle. It eliminates the need for an initial delay after connection and before the first query, as well as the long pauses between queries. It performs as the dongle should have.

**Reply by SwissDani** [reply]
Hi Stefan
Today i cloned your repo to try it with my EMMA.
Unfortunately, I was not successful. In EMMA I activated ModbusTCP without TLS and set the client IP of my laptop. In your code I set the correct local EMMA IP. Unfortunately, I cannot establish a connection. Do you have any idea what else is needed?
My Emma is connected via WIFI and not via LAN. The laptop is in the same WLAN. 
However, the WLAN is not the installer WLAN of EMMA.
I double-checked the IP addresses via the router.

**Reply by StefanOltmann** [reply]
My EMMA is connect through LAN cable, but I’m certain that if must work via WIFI.
It worked with the dongle over WIFI before, so I don’t see why it shouldn’t.

What error message do you get?

**Reply by SwissDani** [reply]
Just a normal java socket exception. Like something is blocked or EMMA is not actually listening on port 502.

**Reply by StefanOltmann** [reply]
Did you verify using `nmap -p 502` that the port is open?

Maybe there is some kind of firewall in between.
On Windows you may try to pause the Windows firewall.

**Reply by SwissDani** [reply]
![image](https://github.com/user-attachments/assets/ee1f892f-7857-4d9a-aa0a-3ba8a93aacaf)
There should be no firewall. Windows Firewall outgoing is open and my laptop and EMMA are in the same subnet.
I will try by LAN next week. Maybe I have more luck.
Thanks for your help.

**Reply by borgqueenx** [reply]
> Sounds great! :)
> 
> Here is code that actually works with EMMA (modified from the previous script): https://github.com/StefanOltmann/emma_modbus_test/blob/main/src/main/kotlin/Main.kt
> 
> The important parts are indeed `slave_id = 0` and setting the client IP address correctly.
> 
> The EMMA is significantly more stable than the dongle. It eliminates the need for an initial delay after connection and before the first query, as well as the long pauses between queries. It performs as the dongle should have.

Thats great news, the wiki should be updated so it includes the emma.

**Reply by StevenHermans** [reply]
@SwissDani , did you get to solve the connection issue? I have a very similar issue: port 502 doesn't 'open'. In the modbus settings on my Emma, I sometimes have 'Connection Successful' but I switches to 'Peer end is disabled' all the time. I specified the port I use to open my homeassistant instance on my raspberry PI IP. Is that the correct one? It's the only one that occasionaly gives the 'Connection successful' state. But even then, the 502 port doesn't open on the Emma.

**Reply by StevenHermans** [reply]
BTW. For what it's worth: my home assistant is running in bridge mode in a docker container but I have the impression it doesn't have any influence... When checking the port 502 on the Emma IP from my home assistant host, it's always showing as closed and I can't get it to 'open'.

**Reply by StevenHermans** [reply]
I got this issue fixed by having  my installer enable the modbus integration from the commissioning procedure. The Management system-1 parameters on Emma are only for Client mode whereas 'server' mode is to be enabled via the commissioning step. Integration now was setup but all the emma entities are in 'unavailable' state

**Comment by onkelandy** [comment]
I'd love to test this, however I'm too stupid. I have connected to EMMAs WiFi to update to the newest firmware. I can make it connect to region003eu5.fusionsolar.huawei.com on port 27250, but cannot find anything about ModbusTCP activation. I only have a user account though, do I need an installer account for making it work?
Port 502 on internal LAN is closed, when doing a port scan 31220 (unknown), 31221 (ftp) and 31240 (ssl) are open. 
When running the Main.kt script I got connection refused of course (or read time out when trying one of the above ports.. no wonder ;))

**Reply by StefanOltmann** [reply]
You need to use the installer access in the FusionSolar mobile app - the web app won’t work.

your installer needs to tell you the password.

If you have the recent firmware you can activate modbus in the settings.

The port will only show up if you explicitly scan for it.

**Reply by m1h5** [reply]
Could you please describe where I can find the ModBus settings in the FusionSolar app? I have installer access, my EMMA is running firmware V100R024C00SPC100 and on my iPhone the most recent version of FusionSolar app. However I just can't find any ModBus settings.
I was told that these settings can be found in Devices -> Emma -> Parameters, but theres nothing ModBus related.

**Reply by tmycgn** [reply]
@m1h5 Check here: https://github.com/wlcrs/huawei_solar/issues/498#issuecomment-2336223081

**Comment by zekolosse** [comment]


**Reply by zekolosse** [reply]
c is all good it works I restarted the whole and I was able to configure the system. thank you very much

**Comment by Geroldin** [comment]
Hi all,
i'm still new to Huawei Solarsystem but been following this integration for some time now.
My solar system was installed this summer with the following setup.
Emma A02 - RS485 (CAN-Bus) - Sun20000-8KTL-M1 Inverter - Directly connected - LUNA20000-10KW-C1 Battery Controller with Luna2000-7-S1 Battery
The Emma is connected via WAN Port to my Home lan and as soon as I get the install codes I should be able to manage the above part of my installation (should be happening next week).
However, in addition I have a Huawei SCharger-22KT-S0 which is connected to the Emma LAN Port to support "PV Power Preferred" charging. 
Are there any plans to integrate the Charger in this constellation? I've read in the "Issues" Section that the Charger will not be supported directly which I can understand. However, since here the Emma is used and it's already supported I was wondering if there are any plans for this?

**Reply by StevenHermans** [reply]
Have you been able to enable the integration for your setup? Did your 502 port 'open'? What type of Homeassistant setup do you have? I have HA setup via a Docker and I also have an Emma/Inverter/battery but I can't get this integration configured. My 502 port on Emma remains closed...I have the most recent firmwares installed and no Dongle. I configure the 'Management system-1 parameters' on my Emma. No luck so far.

**Comment by acsaba1977** [comment]
Hi everyone. I just finish the installation of a new solar system using Huawei LUNA20000 L1 inverter, Battery and backup box having Emma A02. In addition, I have the installer account, so I manage to enable the TCP Modbus on the EMMA device. The problem is same as some of you already mentioned that the port 502 is closed and I cannot read any registry data via the LAN port, nor WAN port. Moreover, SSH has been enabled and the port 22 is open but I haven't manage to connect. I was trying the installer password with users like installer, root etc. without success. Any thoughts would be welcomed.

**Reply by mattia424** [reply]
Dall'app c'è un'opzione per abilitare la comunicazione modbus

**Reply by voyagerft** [reply]
> Dall'app c'è un'opzione per abilitare la comunicazione modbus

È una domanda o un affermazione? Interesserebbe anche a me risolvere questo problema

**Reply by mattia424** [reply]
E' un'affermazione. per far funzionare Modbus TCP su EMMA è sufficiente abilitarlo con l'account installatore. assicurati però di avere l'ultima versione del fw (ad oggi V100R025C00SPC120 funziona correttamente) in quanto versioni precedenti tipo V100R024 non andava ModbusTCP ma solo RTU.

**Comment by keyanbld** [comment]
Bonjour, je cherche actuellement a integrer une Emma-002 a Home Assistant, etant novice je prend toute aide que vous pouvez m'apporter. 
L'installation est composé : 
D'un SUN2000 LC0 gerer par le Emma-002 pour alimenter en partie une borne V2C Trydan. 
Je penses donc a passer par HA pour controller et relever les donnees

**Comment by ruteclrp** [comment]
Please see discussion https://github.com/wlcrs/huawei_solar/issues/491#issuecomment-3274831719
It appears that EMMA is now online in the integration and working even with the SCharger in RO mode.
Especially of interest is the latest MODBUS specification for EMMA in https://github.com/user-attachments/files/22241982/SmartHEMS.V100R025C00SPC102.MODBUS.Interface.Definitions.pdf where the TOU schedule via EMMA is deliberated in the comment field. From my first reading there may be a new element of "effective date" in the TOU period definition? Leave it to experts here to elaborate/confirm?

**Comment by borgqueenx** [comment]
Sad that we cant control the charger using emma, nevertheless thanks everyone for contributing, hopefully a new firmware with better ocpp support soon.
